### PR TITLE
Add tests for io utils to save images

### LINKED ISSUE #5689: Improve test coverage for screenshot watermarking
## 🧰 Task

There were a bunch of uncovered lines in #5494 that it would be good to cover by tests.

@GenevieveBuckley sorry that I didn't comment after @Czaki's latest comment there — it was a busy time. I think he's right that ideally the uncovered lines in the PR (commented by codecov in [the diff view](https://github.com/napari/napari/pull/5494/files)) should be tested.

### PR DESCRIPTION
**Pull request by dalthviz**
# Fixes/Closes

Closes #5689

# Description

Add some tests to increase coverage of lines related with saving viewer screenshots and utils to save images

# References

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue) - add tests

# How has this been tested?

- [x] all tests pass with my change

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: napari/_qt/_tests/test_qt_window.py [hunk 1]
```diff
@@ -103,3 +107,42 @@ def test_menubar_shortcut(make_napari_viewer):
     v.window._toggle_menubar_visible()
     assert not v.window.main_menu.isVisible()
     assert v.window._main_menu_shortcut.isEnabled()
+
+
+@skip_on_win_ci
```
Comment (GenevieveBuckley) [inline napari/_qt/_tests/test_qt_window.py:112]: Why is this being skipped on Windows?

File: napari/_qt/_tests/test_qt_window.py [hunk 2]
```diff
@@ -103,3 +107,42 @@ def test_menubar_shortcut(make_napari_viewer):
...
+def test_screenshot_to_file(make_napari_viewer, tmp_path):
+    """
+    Test taking a screenshot using the Window instance and saving it to a file.
+    """
+    viewer = make_napari_viewer()
+    screenshot_dir_path = tmp_path / "screenshots"
+    screenshot_dir_path.mkdir()
```
Comment (GenevieveBuckley) [inline napari/_qt/_tests/test_qt_window.py]: It seems simpler to not create a 'screenshots' subdirectory, and instead put the "screenshot.png" file directly in the tmp_path folder.

File: napari/utils/_tests/test_io.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+@pytest.mark.parametrize(
+    "image_file", ["image", "image.png", "image.tif", "image.jpg"]
+)
+def test_imsave(tmp_path, image_file):
+    """Save data to image file."""
+    # create image data
+    np.random.seed(0)
+    data = np.random.randint(20, size=(10, 15), dtype=np.ubyte)
+    image_dir_path = tmp_path / "screenshots"
+    image_dir_path.mkdir()
```
Comment (GenevieveBuckley) [inline napari/utils/_tests/test_io.py]: Same as above, could be simpler to put the output files directly in 'tmp_path', instead of creating the extra 'screenshots' subfolder.

File: napari/utils/_tests/test_io.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    image_file_path = image_dir_path / image_file
+    assert not image_file_path.is_file()
+
+    # create image and assert image file creation
+    imsave(str(image_file_path), data)
+    assert image_file_path.is_file()
```
Comment (GenevieveBuckley) [inline napari/utils/_tests/test_io.py:21]: Like your other test, it would also be good to read in the saved image file and compare it with the original data array.

**Review by GenevieveBuckley — CHANGES_REQUESTED** [review]
Thank you for this PR @dalthviz!

My main concern is why one of the tests is skipped on Windows. I'd like to understand that before we merge this. Ideally we'll be able to change things so it does run on Windows too, and at worst we'll add a comment explaining what the restriction is.

File: napari/_qt/_tests/test_qt_window.py (hunk 1 above)
Comment (brisvag) [reply napari/_qt/_tests/test_qt_window.py:112]: All of our screenshot tests are skipped on windows afaik (don't know the history of why, but they *do* fail on windows ci). In fact, this fixture is misnomed, since it's specifically about screenshot tests on windows ci:

```py 
skip_on_win_ci = pytest.mark.skipif(
    sys.platform.startswith('win') and os.getenv('CI', '0') != '0',
    reason='Screenshot tests are not supported on windows CI.',
)
```

Comment (dalthviz) [reply napari/_qt/_tests/test_qt_window.py:112]: The test I used as base for this (https://github.com/napari/napari/blob/0ffc9e827ac0566365f0ec9d440c49e532ae5644/napari/_qt/_tests/test_qt_viewer.py#L225-L229) also has the `skip_on_win_ci`. Since the tests were failing I put it also in the new test. Locally the test passes for me on my Windows machine:

imagen

As mentioned, maybe the `skip_on_win_ci` is not the best name. Maybe the `reason` should be updated to a more generic message and a new alias mark definition should be created with the reason related to the screenshot tests? Maybe the new alias could be something like `skip_screenshots_on_win_ci`?

Comment (GenevieveBuckley) [reply napari/_qt/_tests/test_qt_window.py:112]: Ah, I see. Thank you both for explaining, I agree it's not the best name. I remove my objection 😄

File: napari/_qt/_tests/test_qt_provide_theme.py [hunk 5]
```diff
@@ -22,7 +22,7 @@ def test_provide_theme_hook_registered_correctly(
 ):
     # make a viewer with a plugin & theme registered
     viewer = make_napari_viewer_with_plugin_theme(
-        make_napari_viewer,
+        make_napari_viewer(),
```
Comment (GenevieveBuckley) [inline napari/_qt/_tests/test_qt_provide_theme.py]: This bit is new, can you tell me why it was necessary?
I don't totally understand what is happening here, since this test passes without this change in my local dev environment, and also passes in CI runs from previous PRs.

File: napari/_qt/_tests/test_qt_provide_theme.py [hunk 6]
```diff
@@ -83,10 +83,6 @@ class TestPlugin:
         def napari_experimental_provide_theme():
             return {name: theme}
 
-    # create instance of viewer to make sure
-    # registration and unregistration methods are called
-    viewer = make_napari_viewer()
```
Comment (GenevieveBuckley) [inline napari/_qt/_tests/test_qt_provide_theme.py]: I don't fully understand why this change was made, but I have a bad feeling about it because now it doesn't seem to match the pattern used everywhere else in the napari tests.

```python
def test_something(make_napari_viewer, other_args):
    viewer = make_napari_viewer()
    # ... rest of test...
```

File: napari/_qt/_tests/test_qt_provide_theme.py (hunk 5 above)
Comment (dalthviz) [reply napari/_qt/_tests/test_qt_provide_theme.py]: This change is just me testing if I could get this test to pass since after multiple retries a job kept failing showing a `Windows fatal exception: access violation`: https://github.com/napari/napari/actions/runs/4885713982/jobs/8721688023#step:8:203

I will revert this change :+1:

File: napari/_qt/_tests/test_qt_provide_theme.py (hunk 6 above)
Comment (dalthviz) [reply napari/_qt/_tests/test_qt_provide_theme.py]: Just me testing something. I will revert this!

**Comment by dalthviz** [comment]
Updated the PR with the latest main changes and the job that was failing due to `napari/_qt/_tests/test_qt_provide_theme.py::test_provide_theme_hook_registered_correctly` is now passing :tada: . However, seems like now the macOS job is timing out (https://github.com/napari/napari/actions/runs/4896285141/jobs/8743546138?pr=5780#step:8:273) 🤔 

Should I increase the `timeout-minutes` `test` job config and `Test with tox` step to see if that helps?

File: napari/utils/_tests/test_io.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    # check image content
+    from qtpy.QtGui import QImage
+
+    from napari._qt.utils import QImg2array
+
+    img_to_array = QImg2array(QImage(str(image_file_path)))
+    assert np.equal(data, img_to_array[:, :, 0]).all()
```
Comment (Czaki) [inline napari/utils/_tests/test_io.py]: Why did you use Qt to read instead of imageio?

Comment (dalthviz) [reply napari/utils/_tests/test_io.py]: I just adapted here the code I found when I was checking `qt_mainwindow.py` that uses `QImg2array`. Checking `imageio`, probably I should use [`imageio.v3.imread`](https://imageio.readthedocs.io/en/stable/_autosummary/imageio.v3.imread.html) right?

Comment (Czaki) [reply napari/utils/_tests/test_io.py]: Yes

File: napari/_qt/_tests/test_qt_provide_theme.py (hunk 5 above)
Comment (GenevieveBuckley) [reply napari/_qt/_tests/test_qt_provide_theme.py]: Got it, thank you!

**Review by GenevieveBuckley — APPROVED** [review]

**Comment by GenevieveBuckley** [comment]
When the CI checks pass, this should be ready to merge 🎉

**Comment by GenevieveBuckley** [comment]
Merged! Nice work @dalthviz

**Comment by psobolewskiPhD** [comment]
The tests are related to https://github.com/napari/napari/pull/5494
Which is slated for 0.5.0 so will add the same milestone here.
